### PR TITLE
Add storage.k8s.io/v1 API

### PR DESCRIPTION
**Pull request by jsafrane**
v1 API is direct copy of v1beta1 API. This v1 API gets installed and exposed in this PR, I tested that kubectl can create both v1beta1 and v1 StorageClass.

~~Rest of Kubernetes (controllers, examples,. tests, ...) still use v1beta1 API, I will update it when this PR gets merged as these changes would get lost among generated code.~~ Most parts use v1 API now, it would not compile / run tests without it.

**Release note**:
```
Kubernetes API storage.k8s.io for storage objects is now fully supported and is available as storage.k8s.io/v1. Beta version of the API storage.k8s.io/v1beta1 is still available in this release, however it will be removed in a future Kubernetes release.

Together with the API endpoint, StorageClass annotation "storageclass.beta.kubernetes.io/is-default-class" is deprecated and  "storageclass.kubernetes.io/is-default-class" should be used instead to mark a default storage class. The beta annotation is still working in this release, however it won't be supported in the next one.
```

@kubernetes/sig-storage-misc

### CODE REVIEW

**Comment by jsafrane** [comment]
Rebased (that was fast!) and regenerated, although I do not like newly created files like pkg/apis/meta/v1/zz_generated.deepcopy.go.

**Comment by jsafrane** [comment]
And another rebase in 5 minutes

**Comment by jsafrane** [comment]
I give up, too many changes are made in parallel to this PR. I'll resume my work tomorrow, hoping it settles down a bit. Sorry for the noise.

**Comment by msau42** [comment]
Do you want to consider issue #35874 for the API layout?

**Comment by thockin** [comment]
I don't think we need that.  I ACK the consistency argument, but I don't
think that outweighs the cost of change at this point.  Open to
discussion...

**Comment by jsafrane** [comment]
I'd leave the API as it is, I don't think we would need StorageClass.Status anytime soon.

**Comment by jsafrane** [comment]
2nd attempt, cross your fingers...

**Comment by jsafrane** [comment]
```
errors in package "k8s.io/kubernetes/pkg/client/clientset_generated/clientset/typed/storage/v1/fake":
output for "fake/fake_storageclass.go" differs; first existing/expected diff: 
  "api_v1 \"k8s.io/kubernetes/pkg/api/v1\"\n\tv1 \"k8s.io/kubernetes/pkg/apis/storage/v1\"\n\tcore \"k8s.io/kube"
  "v1 \"k8s.io/kubernetes/pkg/apis/storage/v1\"\n\tcore \"k8s.io/kubernetes/pkg/client/testing/core\"\n)\n\n// F"

I0125 02:12:50.077] FAILED   hack/make-rules/../../hack/verify-codegen.sh	39s
```

Errors in generated code? Really? This is sooo annoying... update-all.sh generates something else every time I do a rebase.

**Comment by jsafrane** [comment]
rebased, cross your fingers...

**Comment by jsafrane** [comment]
oops, wrong PR, no lgtm

**Comment by jsafrane** [comment]
cc @kerneltime, this is part of StorageClass GA effort

**Comment by jsafrane** [comment]
Added a release note, please review (and fix) my English.

**Comment by jsafrane** [comment]
For reviewers: the reason that so many files have to be changed is that generated client now returns v1 StorageClasses instead of v1beta1 ones. This must be reflected in the code (mostly in imports).

File: pkg/apis/storage/v1/util/helpers.go [hunk 1]
```diff
@@ -0,0 +1,137 @@
...
+// It currently matches the Beta value and can change when official is set.
+// - in PersistentVolumeClaim it represents required class to match.
+//   Only PersistentVolumes with the same class (i.e. annotation with the same
+//   value) can be bound to the claim. In case no such volume exists, the
+//   controller will provision a new one using StorageClass instance with
+//   the same name as the annotation value.
+// - in PersistentVolume it represents storage class to which the persistent
+//   volume belongs.
+//TODO: Update this to final annotation value as it matches BetaStorageClassAnnotation for now
+const StorageClassAnnotation = "volume.beta.kubernetes.io/storage-class"
```
Comment (msau42) [inline pkg/apis/storage/v1/util/helpers.go]: Is an annotation still needed in the final version?  Wouldn't the storageclass be specified as an attribute in the PV/PVC spec instead?

Comment (jsafrane) [inline pkg/apis/storage/v1/util/helpers.go]: whole `storage/*/util/helpers.go` is unfortunate, there is v1beta1 version, v1 version and "internal" unversion variant, all with the same annotation definitions. I will refactor that out in a subsequent PRs and leave only version-specific code there. In the end, only `BetaStorageClassAnnotation` will remain to keep backward compatibility for couple of releases.

**Comment by bgrant0607** [comment]
@jsafrane Why does StorageClass need to go GA now?

**Comment by jsafrane** [comment]
>Why does StorageClass need to go GA now?

@bgrant, because it has been in beta for 2 releases and no serious issues were found and it has been decided in sig-storage as 1.6 priority, see https://docs.google.com/spreadsheets/d/1t4z5DYKjX2ZDlkTpCnp18icRAQqOE85C1T1r2gqJVck/edit#gid=0

**Comment by jsafrane** [comment]
Rebased

**Comment by thockin** [comment]
Also, if 1.6 is cleanup release, getting things out of Beta seems useful?

**Comment by jsafrane** [comment]
```
W0214 06:26:54.788] Some packages in hack/.linted_packages do not exist anymore. Please remove them
W0214 06:26:54.788] from hack/.linted_packages:
W0214 06:26:54.789] 
W0214 06:26:54.789]   pkg/client/informers/informers_generated/storage/v1
```

That's very suspicious, I'll check where it comes from and how to regenerate it.

**Comment by jsafrane** [comment]
Fixed verify-golint

**Comment by jsafrane** [comment]
Rebased, sorry about the noise with failed tests.

**Comment by jsafrane** [comment]
assigning to @msau42 for review

**Comment by jsafrane** [comment]
@k8s-bot gce etcd3 e2e test this

**Comment by jsafrane** [comment]
@k8s-bot gce etcd3 e2e test this

File: pkg/apis/storage/v1/util/helpers.go [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+)
+
+// IsDefaultStorageClassAnnotation represents a StorageClass annotation that
+// marks a class as the default StorageClass
+//TODO: Update IsDefaultStorageClassannotation and remove Beta when no longer used
+const IsDefaultStorageClassAnnotation = "storageclass.beta.kubernetes.io/is-default-class"
+const BetaIsDefaultStorageClassAnnotation = "storageclass.beta.kubernetes.io/is-default-class"
+
+// AlphaStorageClassAnnotation represents the previous alpha storage class
+// annotation.  it's no longer used and held here for posterity.
```
Comment (thockin) [inline pkg/apis/storage/v1/util/helpers.go]: Worth removing alphas now?  Or is that a different PR?

**Comment by jsafrane** [comment]
@k8s-bot kubemark e2e test this

**Comment by jsafrane** [comment]
restoring lgtm after rebase

**Comment by jsafrane** [comment]
rebased on top of #40013, cross your fingers...

**Comment by jsafrane** [comment]
@k8s-bot gce etcd3 e2e test this

**Comment by jsafrane** [comment]
@k8s-bot kops aws e2e test this

**Comment by jsafrane** [comment]
@k8s-bot kubemark e2e test this

**Comment by jsafrane** [comment]
@k8s-bot non-cri e2e test this

**Comment by jsafrane** [comment]
@k8s-bot kops aws e2e test this

**Comment by jsafrane** [comment]
@k8s-bot non-cri e2e test this

**Comment by jsafrane** [comment]
Another rebase... This is so frustrating.

**Comment by jsafrane** [comment]
@k8s-bot kops aws e2e test this

**Comment by liggitt** [comment]
It's likely this broke GKE, which explicitly enables API versions and often requires coordination to drop new API versions

**Comment by msau42** [comment]
@liggitt I can coordinate with you to enable the API in GKE if needed.

**Comment by liggitt** [comment]
@msau42 We should enable it if possible.

**Comment by liggitt** [comment]
@k8s-oncall if we want the merge queue to drain over the weekend, we enable storage.k8s.io/v1 in the GKE CI or revert this. I'd prefer getting GKE working, but opened https://github.com/kubernetes/kubernetes/pull/42114 in case we need it.

cc @jsafrane

**Comment by msau42** [comment]
Sorry @jsafrane, this had to be reverted to unblock the merge queue.  I'm working on enabling this in GKE ASAP.  The change is simple so I should be able to check it in today.  But still, it needs to be rolled out, which can take some time.  I'm checking if we can expedite the rollout.

**Comment by msau42** [comment]
Quick update:
I created a dev gke cluster using my GKE changes + this PR.  Then I ran the dynamic provisioning e2e tests that had failed earlier, and they passed.

I just checked in the GKE changes, and I've been told that they rollout to Jenkins every 6 hours.  So hopefully within a few hours from now, it will be active and we can resubmit this PR.

**Comment by msau42** [comment]
Actually, I had to revert the GKE change because the API group needs to exist first.  So we'll need to stage this in 3 parts.  1) Add the new API group  2) Enable it in GKE 3) Change the callers to use the new API group.

File: pkg/apis/storage/v1/util/helpers.go (hunk 1 above)
Comment (r2d4) [inline pkg/apis/storage/v1/util/helpers.go]: Sorry to ping an old issue here - but I'm wondering what the resolution of this was?  The release-note says that we should be using `storageclass.kubernetes.io/` but the code doesn't actually look for that annotation, only the alpha and beta.

Comment (jsafrane) [inline pkg/apis/storage/v1/util/helpers.go]: Do not mix  `volume.beta.kubernetes.io/storage-class` and `volume.beta.kubernetes.io/is-default-class`, these are two annotations with completely different meaning. `is-default-class` was never alpha, only beta. On the other side, you are right, it should be without `beta` in 1.6 as the release note suggests. I filled #42991. Thanks for heads-up!

**Comment by thockin** [comment]
@danwinship an example adding a new apigroup.  Sorry for what I am about to drop on you, this is not as easy as it needs to be.

**Pull request merged**